### DISCUSSION TITLE
How to set custom formatter for ElasticsearchTaskHandler?

### DISCUSSION THREAD

**Discussion by Romsik788** [apache/airflow#28629] [Q&A]
Hi,
I have an Airflow 2.3.0 installation that is configured to be remotely logged into elasticsearch. Here is some part of my env:
```
AIRFLOW__ELASTICSEARCH__LOG_ID_TEMPLATE: "airflow-{dag_id}-{task_id}-{execution_date}-{try_number}"
AIRFLOW__ELASTICSEARCH__WRITE_STDOUT: "true"
AIRFLOW__ELASTICSEARCH__JSON_FORMAT: "true"
AIRFLOW__ELASTICSEARCH__JSON_FIELDS: "asctime, filename, lineno, levelname, message, extra"
AIRFLOW__ELASTICSEARCH__HOST: "192.168.0.2:9200"
AIRFLOW__LOGGING__REMOTE_LOGGING: "true"
```

I have extra fields in the logger, for example:
`self.__logger.warning("Event is invalid", extra={"event": data, "detail": str(e)})`
And I need to write extra fields under one value - **extra**, for example:
```
{"asctime": "2022-12-28, 19:24:04 EET", "filename": "extractor.py", "lineno": 61, "levelname": "CRITICAL", "message": "192.168.0.2 is not available", "extra": "{'detail': ConnectionError(MaxRetryError(\"HTTPConnectionPool(host='192.168.0.2', port=8000): Max retries exceeded with url: /v1/events (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f4b83ef2410>: Failed to establish a new connection: [Errno 111] Connection refused'))\"))}"}
```
For this I decided to use my own class for formatting:
```
EXTRA_KEYS = ["host", "type", "line", "detail", "event", "fname"]

class CustomFormatter(JSONFormatter):
    def __init__(
        self, fmt=None, datefmt=None, style="%", json_fields=None, extras=None
    ):
        super().__init__(fmt, datefmt, style, json_fields, extras)

    def format(self, record):
        extra = {k: v for k, v in record.__dict__.items() if k in EXTRA_KEYS}
        if extra:
            record.__dict__["extra"] = str(extra)
        string = super().format(record)
        return string
```
I also have a function with which I get the logger at the beginning of the dag:
```
JSON_FIELDS = ["asctime", "filename", "lineno", "levelname", "message", "extra"]
def get_logger() -> logging.Logger:
    logger = logging.getLogger("airflow.task")
    logger.handlers[0].setFormatter(
        CustomFormatter(json_fields=JSON_FIELDS)
    )
    return logger
```
But for some reason, I cannot replace the standard formatter with my own in this way. But when I use local logging, it works for FileTaskHandler.
Could you please explain why it works this way? Is there an easy way to replace the default formatter with a CustomFormatter? And if not, then there are simple solutions how to handle my fields other than the obvious one:
`extra={"extra": {"event": data, "detail": str(e)}}`?

**Comment by potiuk** [comment]
You have not explained what does it mean "I cannot replace the standard formatter". I think what would be helpful is if you describe what errors you have and what your debugging brings.

I thinkl your get_logger() in DAG ( which I alos od not understand how it is done). You should define and use your custom logging in the way described in https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/logging-tasks.html#advanced-configuration if you want to add custom formatter.

And I recommend you simply debug it with debugger. With logigng config and especially ES handler there might be subtle problems that you will experience.
